src/crypto/CHIPCryptoPAL.h

```diff
@@ -214,6 +262,426 @@ CHIP_ERROR add_entropy_source(entropy_source fn_source, void * p_source, size_t
+     * @param out     Output buffer
+     * @param out_len Length of the output buffer
+     *
+     *  @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR PointWrite(const void * R, unsigned char * out, size_t out_len) = 0;
+
+    /**
+     * @brief Scalar multiplication, R = fe1 * P1.
+     *
+     * @param R   Resultant point
+     * @param P1  Input point
+     * @param fe1  Input field element.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR PointMul(void * R, const void * P1, const void * fe1) = 0;
+
+    /**
+     * @brief Scalar multiplication with addition, R = fe1 * P1 + fe2 * P2.
+     *
+     * @param R   Resultant point
+     * @param P1  Input point
+     * @param fe1  Input field element.
+     * @param P2  Input point
+     * @param fe2  Input field element.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR PointAddMul(void * R, const void * P1, const void * fe1, const void * P2, const void * fe2) = 0;
+
+    /**
+     * @brief Point inversion.
+     *
+     * @param R   Input/Output point to point_invert
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR PointInvert(void * R) = 0;
+
+    /**
+     * @brief Multiply a point by the curve cofactor.
+     *
+     * @param R   Input/Output point to point_invert
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR PointCofactorMul(void * R) = 0;
+
+    /*
+     *   @synopsis Check if a point is on the curve.
+     *
+     *   @param R   Input point to check.
+     *
+     *   @return CHIP_NO_ERROR if the point is valid, CHIP_ERROR otherwise.
+     */
+    virtual CHIP_ERROR PointIsValid(void * R) = 0;
+
+    /*
+     *   @synopsis Compute w1in*G
+     *
+     *   @param Lout        Output point in 0x04 || X || Y format.
+     *   @param L_len       Output point length
+     *   @param w1in        Input field element
+     *   @param w1in_len    Input field element size
+     *
+     *   @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR ComputeL(unsigned char * Lout, size_t * L_len, const unsigned char * w1in, size_t w1in_len) = 0;
+
+    void * M;
+    void * N;
+    void * G;
+    void * X;
+    void * Y;
+    void * L;
+    void * Z;
+    void * V;
+    void * w0;
+    void * w1;
+    void * xy;
+    void * order;
+    void * tempbn;
+};
+
+#if CHIP_CRYPTO_OPENSSL
+struct openssl_spake2_ctx
+{
+    EC_GROUP * curve;
+    BN_CTX * bn_ctx;
+    EVP_MD_CTX * hash_ctx;
+    const EVP_MD * hash;
+};
+#endif
+
+class Spake2p_P256_SHA256_HKDF_HMAC : public Spake2p
+{
+public:
+    Spake2p_P256_SHA256_HKDF_HMAC(void) : Spake2p(kP256_FE_Length, kP256_Point_Length, kSHA256_Hash_Length) {}
+    virtual ~Spake2p_P256_SHA256_HKDF_HMAC(void) { FreeImpl(); }
```

Comment:
You're calling a virtual function during a destructor. I think it is actually OK here, but some consider it bad practice (see https://www.artima.com/cppsource/nevercall.html). Maybe there's no reason for FreeImpl to be defined as (pure) virtual in the base class ?